src/client/Microsoft.Identity.Client/Region/RegionDiscovery.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Internal.Logger;
+using Microsoft.Identity.Client.PlatformsCommon.Shared;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.Region
+{
+    internal sealed class RegionDiscovery
+    {
+        private const string RegionName = "REGION_NAME";
+        private Uri ImdsUri;
+        private IDictionary<string, string> Headers;
+        private ICoreLogger Logger;
+        private static readonly object _lock = new object();
+        private static RegionDiscovery Instance = null;
+        private HttpManager HttpManager;
+
+        private RegionDiscovery()
+        {
+            ImdsUri = new Uri("http://169.254.169.254/metadata/instance/compute");
+            Headers = new Dictionary<string, string>();
+            Headers.Add("api-version", "2019-06-01");
+            Logger = new NullLogger();
+            HttpManager = new HttpManager(new SimpleHttpClientFactory());
+        }
+
+        // For Unit tests only.
```

Comment:
For unit testing, we have the ability to mock all services from the bundle.

Reply:
But in general I prefer to have 2 constructors - one for product code and one for test code, smth like 

```csharp
public ClassUnderTest(IFoo fooDependnecyForTest, IBar barDepedencyForTest = null)
{
if (fooDependnecyForTest==null) { fooDependnecyForTest = new ... };
}
```

This approach allows you to keep the object immutable. And we like immutable, because [immutable is thread safe](http://tutorials.jenkov.com/java-concurrency/thread-safety-and-immutability.html) and keeps complexity down as well.

Reply:
Changed